### PR TITLE
CORE-12552 Follow up from CORE-10576

### PR DESCRIPTION
**Pull request by williamvigorr3**
This change contains the following:

Functional changes:

1. Change the Member Synchronisation Service to do a look up in known members if the member is not found in the update package. Use the status in the same way, as if it was in the membership package to calculate the expected hash.
2. Add logging to ActivateMemberHandler and SuspendMemberHandler.
3. Use a stream instead of a list in the large queries (for notary updates the group parameters).

Testing coverage

1. Improve unit test coverage of the MGM Synchronisation service. Add coverage of notary suspension/activation to SuspendMemberHandlerTest and ActivateMemberHandlerTest.
2. Test notary activation in the MembershipPersistenceTest.
3. Unit test changes from CORE-10576 in the MgmSynchronisation service.
4. Move coverage of exceptions from the AddNotaryToGroupParameterHandlerTest to the GroupParametersNotaryUpdaterTest. To make purer unit tests.

I left this unaddressed on the original PR: https://github.com/corda/corda-runtime-os/pull/3565#discussion_r1159678991 As this is the only way to return an error to the user in that case.

### CODE REVIEW

File: components/membership/membership-persistence-service-impl/src/integrationTest/kotlin/net/corda/membership/impl/persistence/service/MembershipPersistenceTest.kt [hunk 1]
```diff
@@ -1502,6 +1502,112 @@ class MembershipPersistenceTest {
...
+                    ),
+                    CryptoSignatureWithKey(
+                        ByteBuffer.wrap(signatureKey),
+                        ByteBuffer.wrap(signatureContent)
+                    ),
+                    signatureSpec,
+                )
+            )
+        ).execute()
+        assertThat(memberPersistenceResult1).isInstanceOf(MembershipPersistenceResult.Success::class.java)
```
Comment (yift-r3) [inline components/membership/membership-persistence-service-impl/src/integrationTest/kotlin/net/corda/membership/impl/persistence/service/MembershipPersistenceTest.kt]: We can do `getOrThrow` and avoid this assertion.

File: components/membership/membership-persistence-service-impl/src/integrationTest/kotlin/net/corda/membership/impl/persistence/service/MembershipPersistenceTest.kt [hunk 2]
```diff
@@ -1502,6 +1502,112 @@ class MembershipPersistenceTest {
...
+        assertThat(groupParameters!!.entries.filterNot { it.key == MODIFIED_TIME_KEY })
+            .containsExactlyInAnyOrderElementsOf(expectedGroupParameters.associate { it.key to it.value }.entries)
+
+        val persistedMemberInfoEntity = vnodeEmf.createEntityManager().use {
+            it.find(
+                MemberInfoEntity::class.java,
+                MemberInfoEntityPrimaryKey(viewOwningHoldingIdentity.groupId, memberX500Name.toString(), false)
+            )
+        }
+        assertThat(persistedMemberInfoEntity).isNotNull
```
Comment (yift-r3) [inline components/membership/membership-persistence-service-impl/src/integrationTest/kotlin/net/corda/membership/impl/persistence/service/MembershipPersistenceTest.kt]: I don't think we need this; we can do `?.` in the following lines instead.

File: components/membership/membership-persistence-service-impl/src/integrationTest/kotlin/net/corda/membership/impl/persistence/service/MembershipPersistenceTest.kt [hunk 3]
```diff
@@ -1502,6 +1502,112 @@ class MembershipPersistenceTest {
...
+        assertThat(persistedMemberInfoEntity.status).isEqualTo(MEMBER_STATUS_ACTIVE)
+        assertThat(persistedMemberInfoEntity.serialNumber).isEqualTo(2L)
+        with(updatedMemberInfo.mgmContext.toMap()) {
```
Comment (yift-r3) [inline components/membership/membership-persistence-service-impl/src/integrationTest/kotlin/net/corda/membership/impl/persistence/service/MembershipPersistenceTest.kt]: We can also do `updatedMemberInfo.mgmContext.toMap().containsEntry(STATUS, MEMBER_STATUS_ACTIVE).containsEntry(SERIAL, "2")`

File: components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/AddNotaryToGroupParametersHandler.kt [hunk 4]
```diff
@@ -105,7 +106,7 @@ internal class AddNotaryToGroupParametersHandler(
                 it.notaryDetails?.serviceName.toString() == notaryServiceName
             }.map {
                 it.notaryDetails!!.serviceProtocolVersions.toHashSet()
-            }.reduceOrNull { acc, it -> acc.apply { retainAll(it) } } ?: emptySet()
+            }.toList().reduceOrNull { acc, it -> acc.apply { retainAll(it) } } ?: emptySet()
```
Comment (yift-r3) [inline components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/AddNotaryToGroupParametersHandler.kt]: Maybe `asSequence` instead?

File: components/membership/membership-persistence-service-impl/src/test/kotlin/net/corda/membership/impl/persistence/service/handler/ActivateMemberHandlerTest.kt [hunk 5]
```diff
@@ -116,54 +146,47 @@ class ActivateMemberHandlerTest {
...
-        )
+        assertThat(result.memberInfo).isEqualTo(persistentMemberInfo)
+        assertThat(result.groupParameters).isEqualTo(groupParameters)
+    }
+
+    @Test
+    fun `invoke throws a MembershipPersistenceException if the mgmContext can't be deserialized`() {
+        whenever(keyValuePairListDeserializer.deserialize(serializedMgmContext)).thenReturn(null)
+        assertThrows<MembershipPersistenceException> {  invokeTestFunction() }.also {
+            it.message!!.contains("Failed to deserialize")
```
Comment (yift-r3) [inline components/membership/membership-persistence-service-impl/src/test/kotlin/net/corda/membership/impl/persistence/service/handler/ActivateMemberHandlerTest.kt]: `assertThat(it).hasMessageContaining("Failed to deserialize")` (I think the result of the contains will be ignored)

File: components/membership/membership-persistence-service-impl/src/test/kotlin/net/corda/membership/impl/persistence/service/handler/SuspendMemberHandlerTest.kt [hunk 6]
```diff
@@ -36,30 +36,105 @@ import javax.persistence.EntityManager
...
-    )
-    private val contextBytes = byteArrayOf(1, 10)
     private val holdingIdentity = HoldingIdentity(knownX500Name, knownGroupId)
     private val ourVirtualNodeInfo: VirtualNodeInfo = mock {
         on { vaultDmlConnectionId } doReturn UUID(0, 1)
     }
     private val clock: Clock = TestClock(Instant.ofEpochSecond(1))
 
     private val transaction: EntityTransaction = mock()
+    private val resultList: List<GroupParametersEntity> = mock {
```
Comment (yift-r3) [inline components/membership/membership-persistence-service-impl/src/test/kotlin/net/corda/membership/impl/persistence/service/handler/SuspendMemberHandlerTest.kt]: Why not just `listOf(mock())`?

File: components/membership/membership-persistence-service-impl/src/test/kotlin/net/corda/membership/impl/persistence/service/handler/ActivateMemberHandlerTest.kt [hunk 7]
```diff
@@ -74,10 +76,38 @@ class ActivateMemberHandlerTest {
...
+        on { mgmContext } doReturn serializedMgmContext
+    }
+    private val mgmContext = KeyValuePairList(
+        listOf(
+            KeyValuePair("one", "1")
+        )
+    )
+    private val persistentMemberInfo = mock<PersistentMemberInfo>()
+    private val suspensionActivationEntityOperations =
+        Mockito.mockConstruction(SuspensionActivationEntityOperations::class.java) { mock, _ ->
```
Comment (yift-r3) [inline components/membership/membership-persistence-service-impl/src/test/kotlin/net/corda/membership/impl/persistence/service/handler/ActivateMemberHandlerTest.kt]: We can add the SuspensionActivationEntityOperations to the constructor and avoid the mock construction. (also for the AddNotaryToGroupParametersHandler in ActivateMemberHandler)

File: libs/membership/membership-common/src/test/kotlin/net/corda/membership/lib/GroupParametersNotaryUpdaterTest.kt [hunk 8]
```diff
@@ -72,6 +74,26 @@ class GroupParametersNotaryUpdaterTest {
...
     }
 
+    @Test
+    fun `notary protocol must be specified to add new notary service`() {
+        val publicKey = mock<PublicKey>()
+        val notaryKey = MemberNotaryKey(publicKey, mock(), mock())
+        val notaryToAdd = MemberNotaryDetails(notaryAx500Name, null, setOf(1, 3), listOf(notaryKey))
+
+        val ex = assertThrows<MembershipPersistenceException> { notaryUpdater.addNewNotaryService(originalGroupParameters, notaryToAdd) }
+        assertThat(ex.message).contains("protocol must be specified")
```
Comment (yift-r3) [inline libs/membership/membership-common/src/test/kotlin/net/corda/membership/lib/GroupParametersNotaryUpdaterTest.kt]: `hasMessageContaining`

File: components/membership/membership-persistence-service-impl/src/test/kotlin/net/corda/membership/impl/persistence/service/handler/ActivateMemberHandlerTest.kt (hunk 5 above)
Comment (williamvigorr3) [reply components/membership/membership-persistence-service-impl/src/test/kotlin/net/corda/membership/impl/persistence/service/handler/ActivateMemberHandlerTest.kt]: Good point.

**Review by yift-r3 — APPROVED** [review]

**Pull request merged**